jax/experimental/array_serialization/serialization.py

```diff
@@ -185,6 +187,20 @@ async def release_bytes(self, requested_bytes):
       self._cv.notify_all()
 
 
+async def transfer_shard_to_host(shard: array.Shard) -> np.ndarray:
+  data = shard.data
+  has_pinned_host = any(
+      m.kind == "pinned_host" for m in shard.device.addressable_memories())
+  if config.enable_memories.value and has_pinned_host:
+    # If available, transfer to pinned host memory
+    sharding = jax.sharding.SingleDeviceSharding(shard.device,
+        memory_kind="pinned_host")
+    data = jax.jit(lambda x: x, out_shardings=sharding)(data)
```

Comment:
Use `jax.device_put` instead?

Reply:
Unfortunately doesn't work yet with `memory_type="pinned_host"` and based on our attempts we believe this will require a longer tail of fixes to XLA. @jaro-sevcik can elaborate.

Reply:
It should! We have tests in memories_test.py that shows it does work. It would be nice to not run a xla computation for a transfer of this kind.

Reply:
We need couple more patches in XLA for non-jitted device_put to work: first https://github.com/openxla/xla/pull/14089 (already submitted for review), and then the last commit from 
https://github.com/jaro-sevcik/xla/tree/device-put-memory-kind-sharding (will submit once https://github.com/openxla/xla/pull/14089 lands).

Reply:
Switched over to using `device_put` in this PR.

Reply:
For completeness, this PR now depends on the XLA PR https://github.com/openxla/xla/pull/14268 (that enables copying buffers to a different memory space).